scripts/build-done-parser.js

```diff
@@ -0,0 +1,218 @@
+"use strict"
+
+const {sequence, either, suffix, capture, avoid} = require("compose-regexp")
+const oneOrMore = suffix("+")
+const zeroOrMore = suffix("*")
+const maybe = suffix("?")
+
+const any = /[^]/
+const space = /\s/
+
+const multiLineComment = sequence("/*", suffix("*?", any), "*/")
+
+// |$ not needed at the end since this is the
+// start of a function that has one parameter
+// and we don't scan past said parameter
+// there is necessrily more text following
+// a comment that occurs in this scenario
+const singleLineComment = sequence("//", zeroOrMore(/[^\n]/), "\n")
+
+const ignorable = suffix("*", either(space, multiLineComment, singleLineComment))
+
+// very loose definitions here, knowing that we're matching valid JS
+// space, '(' for the start of the args, '/' for a comment
+const funcName = /[^\s(\/]+/
+// space, '[' and '{' for destructuring, ')' for the end of args, ',' for next argument,
+// '=' for => and '/' for comments
+const argName = /[^\s{[),=\/]+/
+
+const prologue = (____) => maybe(
+	maybe("function", ____, maybe(/\b/, funcName, ____)),
+	"(",
+	____
+)
+
+// This captures the first identifier after skipping
+// (?:(?:function identifier?)? \()?
+const doneMatcher = sequence(/^/, prologue(ignorable), capture(argName))
+
+console.log("without comments: ", sequence(/^/, prologue(/ */), capture(argName)))
+
+// ------------------------------------------------------------ //
+// tests and output if all green
+
+const o = require('ospec')
+
+const test = (input) => (doneMatcher.exec(input)||[]).pop()
+
+o("done parser", () => {
```

Comment:
This belongs with the rest of the tests - let's just put it there. Let's not pollute the build scripts with it. Of course, it'll need modified so it's in the shape of a test, but we already have a few similar tests anyways. And if you need to detect specific line endings, spaces, and such, you can pass `eval("(function\r\n(done) { ... })")` and such for the body - it's pretty straightforward.

If it's wrong, we'll see it on Travis anyways, and you can debug it easily enough with .

Reply:
It's actually more ergonomic like this. You know immediately if the regexp is good to use in the lib without having to do copy/paste operations. The copy/paste is necessary unless we either use a more involved build process (think rollup), but I'd rather not do that now.

I've kept a few `done` detection tests in the suite as well for integration purposes, these are unit tests, and we can't fit them in the api test suite unless we expose the parser publicly.

You can add more integration tests if you want.

One thing I still have to add to the parser test suite is the non-ascii identifiers.

Reply:
I'm still hesitant about the ospec dependency BTW.

Reply:
What do you have against it?

Reply:
Just the idea of us having a circular dependency. I'd rather avoid it specifically *because* of the issue with maintaining LKGRs and such - ospec is rather simple to maintain and I'd like to keep it that way. It's not like TypeScript or anything like that.

Reply:
Also, for the little that this needs, reporting it could be as simple as this:

```js
const failed = []
const scripts = [
  ["function(done)", "done"],
  // and other tests...
]

for (const [source, expect] of scripts) {
  const found = test(source)
  if (found !== expect) {
    failed.push(`Parse ${JSON.stringify(source)}, expected ${expect}, found ${found}`)
  }
}

if (failed.length) {
  console.log("Test errors:")
  for (const line of failed) console.log(line)
} else {
  console.log(`Paste this:\n\n${doneMatcher}`)
}
```

Not sure we need all of ospec to just do that.

Reply:
That's what I had initially :-)

I'll recapitulate the thinking:

1) I want `ospec` version N to have a *dev* dependency on `ospec` version N-1, for its *main* test suite. Depending on a stable `ospec` version for the main test suite will let us introduce breaking changes, if we want, and hack more generally, without breaking our test harness.
This isn't in any way circular (since it's pointing to its past self, and `devDependecy` means it won't recurse), and users won't even see it. Even if we use `"^x,y,z"` versions rather than an exact version as I've done here, we'll end up with the last published version that matches while developing the next version.  You must `--force` `npm` to add the dependency though, but IIUC in this scenario they are being overly restrictive.

2) I used this as a small scale trial, to see if anything would break for some obscure reason I didn't think of. It didn't. So I went ahead and ported the test suite in #17, and it works. I can switch the tests of the `done` parser back to the plain tester function, having it there was not the end goal.